File: exercise/jianzhi_offer/61SerializeATree.js
const { TreeNode } = require('../TreeNode');
var serializeCallback = function(root, arr) {
    if (root === null) {
    	arr.push("null");
    } else {
    	arr.push(root.val);
	    serializeCallback(root.left, arr);
	    serializeCallback(root.right, arr);
    }
    // return arr;
}
// 在js中数组是引用类型
var serialize = function(root) {
	var arr = [];
	serializeCallback(root,arr);
    return arr.join();
};

var deserializeCallback = function(arr) {
	if (arr.length === 0 || arr.length === 1 && arr[0] === 'null') return null;
	var val = arr.shift();
	var root = null;
	if (!Number.isNaN(Number(val))) {
		root = new TreeNode(val);
		root.left = deserializeCallback(arr);
		root.right = deserializeCallback(arr);
	}
	return root;
}

var deserialize = function(data) {
	var arr = data.split(',');
	// console.log(arr);
	return deserializeCallback(arr); 
    
};
var root = new TreeNode(2);
var node1 = new TreeNode(3);
var node2 = new TreeNode(4);
var node3 = new TreeNode(7);
var node4 = new TreeNode(9);
root.left = node1;
root.right = node2;
node1.left = node3;
node2.left = node4;
var s = serialize(root);
console.log(s);
const r = deserialize(s);
console.log(r);

console.log({a:1,a:1})

File: exercise/jianzhi_offer/61SerializeATree/61SerializeATree.py
## 战胜13%
class TreeNode:
	def __init__(self, val):
		self.val = val
		self.left = None
		self.right = None

class SerializeATree:

    def serializeCallback(self, root, arr):
        if root == None:
            arr.append("null")
        else:
            arr.append(root.val)
            self.serializeCallback(root.left, arr)
            self.serializeCallback(root.right, arr)

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        arr = []
        self.serializeCallback(root, arr)
        return ",".join(str(i) for i in arr)
        
    def deserializeCallback(self, arr):
        if len(arr) == 0 or len(arr) == 1 and arr[0] == "null":
            return None
        value = arr.pop(0)
        node = None
        if value != "null":
            print(value)
            node = TreeNode(value)
            node.left = self.deserializeCallback(arr)
            node.right = self.deserializeCallback(arr)
        return node


    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        arr = data.split(",")
        return self.deserializeCallback(arr)

if __name__ == '__main__':
    root = TreeNode(1)
    node2 = TreeNode(2)
    node3 = TreeNode(3)
    node4 = TreeNode(4)
    node5 = TreeNode(5)
    root.left = node2
    root.right = node3
    node3.left = node4
    node3.right = node5
    serial = SerializeATree()
    str = serial.serialize(root)
    # print('ff')  
    print(str)
    node = serial.deserialize(str)

    print(node)


    

